File: app/src/main/java/com/afterroot/photos/ImageViewerActivity.java
package com.afterroot.photos;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;

import com.afterroot.photos.adapter.ImagePagerAdapter;
import com.afterroot.photos.adapter.ImagesRecyclerViewAdapter;
import com.transitionseverywhere.Explode;
import com.transitionseverywhere.Slide;
import com.transitionseverywhere.TransitionManager;

import java.io.File;

import static com.afterroot.photos.Helper.Static.EXTRA_GOTO_RECYCLER_POS;
import static com.afterroot.photos.Helper.Static.EXTRA_VIEWPAGER_POS;

public class ImageViewerActivity extends AppCompatActivity {
    
    private static final int UI_ANIMATION_DELAY = 0;
    private final Handler mHideHandler = new Handler();
    private ViewPager mImagesPager;
    private final Runnable mHideSystemUi = new Runnable() {
        @SuppressLint("InlinedApi")
        @Override
        public void run() {
            mImagesPager.setSystemUiVisibility(
                    View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                            | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                            | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                            | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                            | View.SYSTEM_UI_FLAG_FULLSCREEN
                            | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
        }
    };

    private boolean mVisible;
    private final Runnable mHideRunnable = this::hide;
    private ActionBar mToolbar;
    private LinearLayout bottomBarLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_image_viewer);
        mToolbar = getSupportActionBar();
        if (mToolbar != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        mVisible = true;
        mImagesPager = (ViewPager) findViewById(R.id.images_pager);

        load();
    }

    ImagePagerAdapter mPagerAdapter;
    Helper mHelper;
    public void load(){
        mHelper = new Helper(this);
        mPagerAdapter = new ImagePagerAdapter(this, mHelper.getImagePaths(mHelper.getFilterTag(), mHelper.getSortOrder()),
                new ImagesRecyclerViewAdapter.OnItemInteractionListener() {
            @Override
            public void onItemClick(int pos) {
                toggle();
            }

            @Override
            public boolean onItemLongClick(int pos) {
                return false;
            }
        });

        bottomBarLayout = (LinearLayout) findViewById(R.id.bottom_bar_layout);
        ImageButton shareButton = (ImageButton) findViewById(R.id.share_button);
        shareButton.setOnClickListener(v -> {
            File file = new File(mPagerAdapter.getPath(mImagesPager.getCurrentItem()));
            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("image/*");
            shareIntent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(file));
            Intent chooser = Intent.createChooser(shareIntent, "Share Image");
            startActivity(chooser);
        });

        assert mImagesPager != null;
        mImagesPager.setAdapter(mPagerAdapter);
        //mImagesPager.setPageTransformer(true, new ZoomOutPageTransformer());
        File file = new File(mPagerAdapter.getPath(mImagesPager.getCurrentItem()));
        try {
            getSupportActionBar().setTitle(mHelper.getFilterTag());
        } catch (Exception e){
            e.printStackTrace();
        }
        mImagesPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                Intent i = new Intent();
                i.putExtra(EXTRA_GOTO_RECYCLER_POS, mImagesPager.getCurrentItem());
                File file = new File(mPagerAdapter.getPath(mImagesPager.getCurrentItem()));
                setResult(RESULT_OK, i);
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

        mImagesPager.setCurrentItem(getIntent().getExtras().getInt(EXTRA_VIEWPAGER_POS, 0));
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        delayedHide(0);
    }

    private void toggle() {
        if (mVisible) {
            hide();
        } else {
            show();
        }
    }

    private void hide() {
        if (mToolbar != null) {
            mToolbar.hide();
        }
        if (bottomBarLayout != null){
            TransitionManager.beginDelayedTransition((ViewGroup) findViewById(R.id.image_viewer_root), new Explode());
            bottomBarLayout.setVisibility(View.INVISIBLE);
        }
        mVisible = false;
        mHideHandler.postDelayed(mHideSystemUi, UI_ANIMATION_DELAY);
    }

    @SuppressLint("InlinedApi")
    private void show() {
        mImagesPager.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
        mVisible = true;
        if (mToolbar != null) {
            mToolbar.show();
        }
        if (bottomBarLayout != null){
            TransitionManager.beginDelayedTransition((ViewGroup) findViewById(R.id.image_viewer_root), new Slide(Gravity.BOTTOM));
            bottomBarLayout.setVisibility(View.VISIBLE);
        }
        
        mHideHandler.removeCallbacks(mHideSystemUi);
    }
    
    private void delayedHide(int delayMillis) {
        mHideHandler.removeCallbacks(mHideRunnable);
        mHideHandler.postDelayed(mHideRunnable, delayMillis);
    }

    private class ZoomOutPageTransformer implements ViewPager.PageTransformer {
        private static final float MIN_SCALE = 0.85f;
        private static final float MIN_ALPHA = 0.5f;

        public void transformPage(View view, float position) {
            int pageWidth = view.getWidth();
            int pageHeight = view.getHeight();

            if (position < -1) { // [-Infinity,-1)
                // This page is way off-screen to the left.
                view.setAlpha(0);

            } else if (position <= 1) { // [-1,1]
                // Modify the default slide transition to shrink the page as well
                float scaleFactor = Math.max(MIN_SCALE, 1 - Math.abs(position));
                float vertMargin = pageHeight * (1 - scaleFactor) / 2;
                float horzMargin = pageWidth * (1 - scaleFactor) / 2;
                if (position < 0) {
                    view.setTranslationX(horzMargin - vertMargin / 2);
                } else {
                    view.setTranslationX(-horzMargin + vertMargin / 2);
                }

                // Scale the page down (between MIN_SCALE and 1)
                view.setScaleX(scaleFactor);
                view.setScaleY(scaleFactor);

                // Fade the page relative to its size.
                view.setAlpha(MIN_ALPHA +
                        (scaleFactor - MIN_SCALE) /
                                (1 - MIN_SCALE) * (1 - MIN_ALPHA));

            } else { // (1,+Infinity]
                // This page is way off-screen to the right.
                view.setAlpha(0);
            }
        }
    }
}

File: app/src/main/java/com/afterroot/tagit/ImageViewerActivity.kt
package com.afterroot.tagit

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.support.v4.view.ViewPager
import android.support.v7.app.ActionBar
import android.support.v7.app.AppCompatActivity
import android.util.DisplayMetrics
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.LinearLayout

import com.afterroot.tagit.adapter.ImagePagerAdapter
import com.afterroot.tagit.adapter.ImagesRecyclerViewAdapter
import com.transitionseverywhere.Explode
import com.transitionseverywhere.Slide
import com.transitionseverywhere.TransitionManager

import java.io.File

class ImageViewerActivity : AppCompatActivity() {
    private val mHideHandler = Handler()
    private var mImagesPager: ViewPager? = null
    private val mHideSystemUi = Runnable {
        mImagesPager!!.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or
                View.SYSTEM_UI_FLAG_FULLSCREEN or
                View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY }

    private var mVisible: Boolean = false
    private var mToolbar: ActionBar? = null
    private var bottomBarLayout: LinearLayout? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_image_viewer)
        mToolbar = supportActionBar
        if (mToolbar != null) {
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        }
        mToolbar!!.setBackgroundDrawable(ColorDrawable(resources.getColor(R.color.transparent_black)))

        mVisible = true
        mImagesPager = findViewById<View>(R.id.images_pager) as ViewPager

        load()
    }

    private lateinit var mPagerAdapter: ImagePagerAdapter
    private lateinit var mHelper: Helper
    private fun load() {
        mHelper = Helper(this)
        mPagerAdapter = ImagePagerAdapter(this, mHelper.getImagePaths(mHelper.filterTag, mHelper.sortOrder),
                object : ImagesRecyclerViewAdapter.OnItemInteractionListener {
                    override fun onItemClick(pos: Int) {
                        toggle()
                    }

                    override fun onItemLongClick(pos: Int): Boolean {
                        return false
                    }
                })

        bottomBarLayout = findViewById<View>(R.id.bottom_bar_layout) as LinearLayout
        val shareButton = findViewById<View>(R.id.share_button) as ImageButton
        shareButton.setOnClickListener { v ->
            val file = File(mPagerAdapter.getPath(mImagesPager!!.currentItem))
            val shareIntent = Intent(Intent.ACTION_SEND)
            shareIntent.type = "image/*"
            shareIntent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(file))
            val chooser = Intent.createChooser(shareIntent, "Share Image")
            startActivity(chooser)
        }

        assert(mImagesPager != null)
        mImagesPager!!.adapter = mPagerAdapter
        try {
            supportActionBar!!.title = mHelper.filterTag
        } catch (e: Exception) {
            e.printStackTrace()
        }

        mImagesPager!!.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

            }

            override fun onPageSelected(position: Int) {
                val i = Intent()
                i.putExtra(Helper.EXTRA_GOTO_RECYCLER_POS, mImagesPager!!.currentItem)
                setResult(Activity.RESULT_OK, i)
            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })

        mImagesPager!!.currentItem = intent.extras!!.getInt(Helper.EXTRA_VIEWPAGER_POS, 0)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        hide()
    }

    private fun toggle() {
        if (mVisible) {
            hide()
        } else {
            show()
        }
    }

    private fun hide() {
        if (mToolbar != null) {
            mToolbar!!.hide()
        }
        if (bottomBarLayout != null) {
            TransitionManager.beginDelayedTransition(findViewById<View>(R.id.image_viewer_root) as ViewGroup, Explode())
            bottomBarLayout!!.visibility = View.INVISIBLE
        }
        mVisible = false
        mHideHandler.postDelayed(mHideSystemUi, UI_ANIMATION_DELAY.toLong())
    }

    @SuppressLint("InlinedApi")
    private fun show() {
        mImagesPager!!.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
        mVisible = true
        if (mToolbar != null) {
            mToolbar!!.show()
        }
        if (bottomBarLayout != null) {
            TransitionManager.beginDelayedTransition(findViewById<View>(R.id.image_viewer_root) as ViewGroup, Slide(Gravity.BOTTOM))
            bottomBarLayout!!.visibility = View.VISIBLE
        }

        mHideHandler.removeCallbacks(mHideSystemUi)
    }

    companion object {

        private val UI_ANIMATION_DELAY = 0

        fun getSoftButtonsBarSizePort(activity: Activity): Int {
            // getRealMetrics is only available with API 17 and +
            val metrics = DisplayMetrics()
            activity.windowManager.defaultDisplay.getMetrics(metrics)
            val usableHeight = metrics.heightPixels
            activity.windowManager.defaultDisplay.getRealMetrics(metrics)
            val realHeight = metrics.heightPixels
            return if (realHeight > usableHeight)
                realHeight - usableHeight
            else
                0
        }
    }
}
